README.md
# git-gutter.el [![travis badge][travis-badge]][travis-link] [![melpa badge][melpa-badge]][melpa-link] [![melpa stable badge][melpa-stable-badge]][melpa-stable-link]

## Introduction

`git-gutter.el` is port of [GitGutter](https://github.com/jisaacks/GitGutter)
which is a plugin of Sublime Text.


## Features

- Asynchronous updating
- [Live updating](#live-updating)
- [Support multiple VCS](#backends)
- Support Tramp
- Work without `vc-mode`


## Screenshot

![git-gutter.el](image/git-gutter1.png)


## Requirements

* Emacs 24 or higher
* [Git](http://git-scm.com/)(1.7.0 or higher)


## git-gutter.el vs [git-gutter-fringe.el](https://github.com/syohex/emacs-git-gutter-fringe)

|                      | git-gutter.el   | git-gutter-fringe.el |
|:---------------------|:---------------:|:--------------------:|
| Work in tty frame    | OK              | NG                   |
| Work with linum-mode | OK(experimental)| OK                   |
| Show on right side   | NG              | OK                   |
| More configurable    | OK              | NG                   |


## Installation

You can install `git-gutter.el` from [MELPA](http://melpa.org) with package.el
(`M-x package-install git-gutter`).

And you can also install it with [el-get](https://github.com/dimitri/el-get).


## Global Minor Mode and Minor Mode

`git-gutter.el` provides global minor-mode(`global-git-gutter-mode`) and minor-mode(`git-gutter-mode`).

If you want to use `git-gutter` for files in git repository.
You add following s-exp in your configuration file(`~/.emacs.d/init.el` or `~/.emacs`).

```lisp
(global-git-gutter-mode +1)
```

Other case, you want to use `git-gutter` for some files, you can use `git-gutter-mode`.
Following example of enabling `git-gutter` for some mode.

```lisp
(add-hook 'ruby-mode-hook 'git-gutter-mode)
(add-hook 'python-mode-hook 'git-gutter-mode)
```

## Commands

`git-gutter.el` provides following commands.
**Obsoleted interfaces will be removed when 1.0 released.**

#### `git-gutter:next-hunk`

Jump to next hunk

#### `git-gutter:previous-hunk`

Jump to previous hunk

#### `git-gutter:mark-hunk`

Mark current hunk.

#### `git-gutter:set-start-revision`

Set start revision where got diff(`git diff`, `hg diff` or `bzr diff`) from.

#### `git-gutter:popup-hunk`

Popup current diff hunk(alias `git-gutter:popup-diff`)

`git-gutter:next-hunk` and `git-gutter:previous-hunk` update content
of buffer popuped by `git-gutter:popup-diff` to current hunk.

#### `git-gutter:stage-hunk`

Stage current hunk. You can use this command like `git add -p`.
This command is supported only for `git`.

#### `git-gutter:revert-hunk`

Revert current hunk

#### `git-gutter`

Show changes from last commit or Update change information.
Please execute this command if diff information is not be updated.

#### `git-gutter:linum-setup`

Setup for working with `linum-mode`.

#### `git-gutter:update-all-windows`

Update git-gutter information of buffers in all visible window.


## Sample Configuration

```lisp
(require 'git-gutter)

;; If you enable global minor mode
(global-git-gutter-mode t)

;; If you would like to use git-gutter.el and linum-mode
(git-gutter:linum-setup)

;; If you enable git-gutter-mode for some modes
(add-hook 'ruby-mode-hook 'git-gutter-mode)

(global-set-key (kbd "C-x C-g") 'git-gutter:toggle)
(global-set-key (kbd "C-x v =") 'git-gutter:popup-hunk)

;; Jump to next/previous hunk
(global-set-key (kbd "C-x p") 'git-gutter:previous-hunk)
(global-set-key (kbd "C-x n") 'git-gutter:next-hunk)

;; Stage current hunk
(global-set-key (kbd "C-x v s") 'git-gutter:stage-hunk)

;; Revert current hunk
(global-set-key (kbd "C-x v r") 'git-gutter:revert-hunk)

;; Mark current hunk
(global-set-key (kbd "C-x v SPC") #'git-gutter:mark-hunk)
```


## Customize

### Live updating

If you set `git-gutter:update-interval` seconds larger than 0, `git-gutter` updates
diff information in real-time by idle timer.

```lisp
(custom-set-variables
 '(git-gutter:update-interval 2))
```

You can stop timer by `git-gutter:cancel-update-timer` and starts by `git-gutter:start-update-timer`.

### Look and feel

![git-gutter-multichar](image/git-gutter-multichar.png)

You can change the signs and those faces.

```lisp
(custom-set-variables
 '(git-gutter:modified-sign "  ") ;; two space
 '(git-gutter:added-sign "++")    ;; multiple character is OK
 '(git-gutter:deleted-sign "--"))

(set-face-background 'git-gutter:modified "purple") ;; background color
(set-face-foreground 'git-gutter:added "green")
(set-face-foreground 'git-gutter:deleted "red")
```

You can change minor-mode name in mode-line to set `git-gutter:lighter`.
Default is " GitGutter"

```lisp
;; first character should be a space
(custom-set-variables
 '(git-gutter:lighter " GG"))
```


### Using full width characters

![git-gutter-fullwidth](image/git-gutter-fullwidth.png)

Emacs has `char-width` function which returns character width.
`git-gutter.el` uses it for calculating character length of the signs.
But `char-width` does not work for some full-width characters.
So you should explicitly specify window width, if you use full-width
character.

```lisp
(custom-set-variables
 '(git-gutter:window-width 2)
 '(git-gutter:modified-sign "☁")
 '(git-gutter:added-sign "☀")
 '(git-gutter:deleted-sign "☂"))
```

### Backends

`git-gutter.el` supports following version control systems

- [Git](http://git-scm.com/)(1.7.0 or higher)
- [Mercurial](https://www.mercurial-scm.org/)
- [Subversion](https://subversion.apache.org/)(1.8 or higher)
- [Bazaar](http://bazaar.canonical.com/en/)


You can set backends which `git-gutter.el` will be used.
Default value of `git-gutter:handled-backends` is `'(git)`. If you want to use
`git-gutter.el` for other VCS, please change value of `git-gutter:handled-backends` as below.

```lisp
;; Use for 'Git'(`git`), 'Mercurial'(`hg`), 'Bazaar'(`bzr`), and 'Subversion'(`svn`) projects
(custom-set-variables
 '(git-gutter:handled-backends '(git hg bzr svn)))
```

### Updates hooks

diff information is updated at hooks in `git-gutter:update-hooks`.

```lisp
(add-to-list 'git-gutter:update-hooks 'focus-in-hook)
```

### Updates commands

diff information is updated after command in `git-gutter:update-commands` executed.

```lisp
(add-to-list 'git-gutter:update-commands 'other-window)
```

### Disabled modes

If you use `global-git-gutter-mode`, you may want some modes to disable
`git-gutter-mode`. You can make it by setting `git-gutter:disabled-modes`
to `non-nil`.

```lisp
;; inactivate git-gutter-mode in asm-mode and image-mode
(custom-set-variables
 '(git-gutter:disabled-modes '(asm-mode image-mode)))
```

Default is `nil`.

### Show signs at gutter by visual lines

Emacs folds long line if `truncate-lines` is `nil`. If `git-gutter:visual-line` is
non-nil, `git-gutter` puts sign by visual lines.

```lisp
(custom-set-variables
 '(git-gutter:visual-line t))
```

Default bahavior is that signs are put by logical lines.
value of `git-gutter:visual-line` is `nil`.

### Show Unchanged Information

![git-gutter-unchanged](image/git-gutter-unchanged.png)

`git-gutter.el` can view unchanged information by setting `git-gutter:unchanged-sign`.
Like following.

```lisp
(custom-set-variables
 '(git-gutter:unchanged-sign " "))
(set-face-background 'git-gutter:unchanged "yellow")
```

Default value of `git-gutter:unchanged-sign` is `nil`.

### Show a separator column

![git-gutter-separator](image/git-gutter-separator.png)

`git-gutter.el` can display an additional separator character at the right of the changed
signs. This is mostly useful when running emacs in a console.

```lisp
(custom-set-variables
 '(git-gutter:separator-sign "|"))
(set-face-foreground 'git-gutter:separator "yellow")
```

Default value of `git-gutter:separator-sign` is `nil`.

Please set `git-gutter:always-show-separator` to non-nil, if you want to show
separator always.

### Hide gutter if there are no changes

Hide gutter when there are no changes if `git-gutter:hide-gutter` is non-nil.
(Default is nil)

```lisp
(custom-set-variables
 '(git-gutter:hide-gutter t))
```

### Pass option to 'git diff' command

You can pass `git diff` option to set `git-gutter:diff-option`.

```lisp
;; ignore all spaces
(custom-set-variables
 '(git-gutter:diff-option "-w"))
```

### Don't ask whether commit/revert or not

`git-gutter.el` always asks you whether commit/revert or not. If you don't want,
please set `git-gutter:ask-p` to `nil`.

```lisp
;; Don't ask me!!
(custom-set-variables
 '(git-gutter:ask-p nil))
```

### Log/Message Level

```lisp
;; Don't need log/message.
(custom-set-variables
 '(git-gutter:verbosity 0))
```

Default value is 4(`0` is lowest, `4` is highest).

### Run hook

Run hook `git-gutter-mode-on-hook` when `git-gutter-mode` is turn on, and
run hook `git-gutter-mode-off-hook` when `git-gutter-mode` is turn off.

## Statistic

`git-gutter.el` provides some statistic API. This is useful for knowing how much
code you changed etc. To display them in mode-line is also useful.

#### `(git-gutter:buffer-hunks)`

Cound unstaged hunks in current buffer.

#### `(git-gutter:all-hunks)`

Cound unstaged hunks in all buffers

#### `(git-gutter:statistic)`

Return statistic unstaged hunks in current buffer. Return value is dot-list.
First element is total added lines, second element is total deleted lines.

## See Also

### [GitGutter](https://github.com/jisaacks/GitGutter)

GitGutter is Sublime Text plugin.

### [vim-gitgutter](https://github.com/airblade/vim-gitgutter)

Vim version of GitGutter

### [diff-hl](https://github.com/dgutov/diff-hl)

`diff-hl` is similar tool based on `vc`.

### [git-gutter-plus](https://github.com/nonsequitur/git-gutter-plus)

Fork of `git-gutter.el`. Some features which are not provided `git-gutter.el` provides.
However git-gutter-plus updates diff information synchronously.


### Another implementation of git-gutter.el

[How to write another implementation](wiki/Write-another-git-gutter.el-implementation)

[travis-badge]: https://travis-ci.org/syohex/emacs-git-gutter.svg
[travis-link]: https://travis-ci.org/syohex/emacs-git-gutter
[melpa-link]: http://melpa.org/#/git-gutter
[melpa-stable-link]: http://stable.melpa.org/#/git-gutter
[melpa-badge]: http://melpa.org/packages/git-gutter-badge.svg
[melpa-stable-badge]: http://stable.melpa.org/packages/git-gutter-badge.svg


Changes
Revision history for git-gutter.el

Revision 0.88 2016/05/21
    - Improve performance(#125)
    - Optimize putting signs(#123)
    - Support visual line(#122)
    - Add statistic commands(#118)

Revision 0.87 2016/01/27
    - Fix mixing face issue(#115)

Revision 0.86 2016/01/24
    - Add mark hunk command(#112)
    - Refactoring code
    - Remove old magit hook setting, it is no longer used(#110)
    - Update unchanged line at linum update(#109
    - Don't upgrade index at invoking 'git diff'(#105)

Revision 0.85 2015/12/10
    - Use EVM for travis-ci
    - Update documents
    - Improve older commands based on git-gutter-mode
    - Refactoring

Revision 0.84 2015/11/23
    - Fix staging command issue on Windows(#101)
    - Add ask flag for staging and reverting command(#96)
    - Fix dead buffer issue for live-update(#92)

Revision 0.83 2015/05/01
    - Fix narrow region issue in staging hunk

Revision 0.82 2015/04/01
    - Fix git-gutter command when git-gutter-mode is disabled

Revision 0.81 2015/03/28
    - Implement live update feature
    - Support Subversion(Thanks alpha22jp)
    - Improve log message

Revision 0.80 2014/12/21
    - Add flag variable for always showing separator

Revision 0.79 2014/11/13
    [Incompatible Change]
     - mercurial is not default backend now
       If you use git-gutter.el for mercurial, please set
       'hg' to git-gutter:handled-backends

Revision 0.78 2014/10/25
    - Support bzr
    Thanks nikolas

Revision 0.77 2014/10/08
    - Add update command which updates all visible buffers

Revision 0.76 2014/09/18
    - Fix staging hunk command bug

Revision 0.75 2014/09/14
    - Update gutter information at magit-revert-buffer-hook

Revision 0.74 2014/09/06
   - Log level is 0 as default now.

Revision 0.73 2014/08/27
   - Add margin width validation

Revision 0.72 2014/08/04
   - Improve for backend configuration

Revision 0.71 2014/07/26
   - Implement git-gutter:set-start-revision
     We can see diff information from arbitrary revision

Revision 0.70 2014/07/22
   - Improve updating gutter information after kill-buffer command

Revision 0.69 2014/07/10
   - Suppress warning if git or hg are not installed.
     (Reported by tom-tan)

Revision 0.68 2014/07/10
   - Support mercurial(Experimental)

Revision 0.67 2014/07/09
   - Don't call linum setup if git-gutter-fringe is used.

Revision 0.66 2014/07/08
   - Fix for linum-mode

Revision 0.65 2014/07/05
   - Fix for vc-revert
   - Fix for new helm

Revision 0.64 2014/07/02 syohex
   - Fix typo function-name and add vc-revert to updating commands
     Thanks thierryvolpiatto
   - Improve view
     reduce 'save-excursion' as possible

Revision 0.63 2014/06/29 syohex
   - Set up linum if global-linum-mode is enabled when git-gutter.el is loaded

Revision 0.62 2014/06/23 syohex
   - Support working with linum-mode(Experimental)

Revision 0.61 2014/06/09 syohex
   - Add verify process buffer is alive
     Thanks thomasf

Revision 0.60 2014/06/01 syohex
   - Support asynchronous updating diff information
   - Reduce needless updating diff information as possible
   - Improve for readability
   - Deprecate git-gutter:xxx-diff commands
     Use git-gutter:xxx-hunk commands instead
   - Improve popup command
   - Drop Emacs 23 support
   - Fix MacOSX display issue

Revision 0.57 2014/04/19 syohex
   - Fix case use set read-only-mode to diff-mode-hook

Revision 0.56 2014/04/16 syohex
   - Switch to cl-lib
   - Fix some bugs

Revision 0.55 2014/03/23 syohex
   - Query before staging hunk

Revision 0.54 2014/02/26 syohex
   - Fix invalid remove-hook
     Thanks m2ym!!

Revision 0.53 2014/01/16 syohex
   - Enable view-mode in popup buffer

Revision 0.52 2013/10/02 syohex
   - Support staging hunk

Revision 0.51 2013/10/01 syohex
   - Suppress byte-compile warnings

Revision 0.50 2013/09/30 syohex
   - Fix git-gutter:revert-hunk at first line

Revision 0.49 2013/09/19 syohex
   - Fix ediff-mode issue(Thanks hanabokuro)

Revision 0.48 2013/06/13 syohex
   - Fix typo group name

Revision 0.47 2013/06/13 syohex
   - Fix for byte compiling warning
   - Shell Quote file name in command.

Revision 0.46 2013/05/25 syohex
   - Fix for Mingw Emacs + Cygwin git environment(Thanks "Cornelius Mika")

Revision 0.45 2013/05/25 syohex
   - Refactoring codes(Thanks "Cornelius Mika")

Revision 0.44 2013/04/27 syohex
   - Fix for deleted hunk at line 1

Revision 0.43 2013/04/23 syohex
   - Support indirect buffer(Reported by @ZaneA)

Revision 0.42 2013/04/16 syohex
   - Fix problem of narrow-to-region(Reported by @ZaneA)

Revision 0.41 2013/04/15 syohex
   - Fix tramp problem lower than Emacs 23.3(Reported by @hirose31)
     (Thanks thierryvolpiatto for good advice)

Revision 0.40 2013/04/14 syohex
   - Improve for window-configuration-change-hook.
     (Update function is called too many times at this hook point)

Revision 0.39 2013/03/28 syohex
   - Implement disabled some modes

Revision 0.38 2013/03/27 syohex
   - Added customization for log level verbosity(by thomasf)

Revision 0.37 2013/03/27 syohex
   - Fix bug
     git-gutter:popup-hunk does not work at deleted hunk.

Revision 0.36 2013/03/20 syohex
   - Implement separator charactor(by sdelafond)
     It is useful in no-window Emacs
   - Fix for byte-compiling

Revision 0.35 2013/03/18 syohex
   [INCOMPATIBLE CHANGE]
   - Remove git-gutter:always-show-gutter and introduce git-gutter:hide-gutter
   - Always show gutter if global-git-gutter is enable.
   - Show message at git-gutter:{next,previous}-hunk if there are no changes.

Revision 0.34 2013/03/17 syohex
   - Remove unneed hook
   - Clean up codes
   - Users can customize update points.

Revision 0.33 2013/03/16 syohex
   - Fix toggle flag bug
   - leaf functions are redefined as inline function

Revision 0.32 2013/03/14 syohex
   - Rename git-gutter:popup-diff to git-gutter:popup-hunk
     User can use old function name as alias

Revision 0.31 2013/03/11 syohex
   - Fix git-gutter:always-show-gutter bug

test/test-git-gutter.el
;;; test-git-gutter.el --- Test for git-gutter.el

;; Copyright (C) 2016 by Syohei YOSHIDA

;; Author: Syohei YOSHIDA <syohex@gmail.com>

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'ert)
(require 'git-gutter)

;; suppress log message
(setq git-gutter:verbosity 0)

(ert-deftest git-gutter:sign-width ()
  "helper function `git-gutter:sign-width'"
  (let ((got1 (git-gutter:sign-width "a"))
        (got2 (git-gutter:sign-width "0123456789")))
    (should (= got1 1))
    (should (= got2 10))))

(ert-deftest git-gutter:select-face ()
  "helper function `git-gutter:select-face'"
  (cl-loop for (type . expected) in '((added . git-gutter:added)
                                      (modified . git-gutter:modified)
                                      (deleted . git-gutter:deleted))
           do
           (should (eq (git-gutter:select-face type) expected)))
  (should-not (git-gutter:select-face 'not-found)))

(ert-deftest git-gutter:select-sign ()
  "helper function `git-gutter:select-sign'"
  (cl-loop for (type . expected) in '((added . "+") (modified . "=") (deleted . "-"))
           do
           (should (string= (git-gutter:select-sign type) expected)))
  (should-not (git-gutter:select-sign 'not-found)))

(ert-deftest git-gutter:propertized-sign ()
  "helper function `git-gutter:propertized-sign'"
  (should (string= (git-gutter:propertized-sign 'added) "+")))

(ert-deftest git-gutter:changes-to-number ()
  "helper function `git-gutter:changes-to-number'"
  (should (= (git-gutter:changes-to-number "") 1))
  (should (= (git-gutter:changes-to-number "123") 123)))

(ert-deftest git-gutter:in-git-repository-p ()
  "Should return nil if default-directory does not exist"

  ;; In git repository, but here is '.git'
  (when (file-directory-p ".git") ;; #36
    (let ((buf (find-file-noselect ".git/config")))
      (with-current-buffer buf
        (should-not (git-gutter:in-git-repository-p)))))

  (let ((default-directory (file-name-directory (locate-library "git-gutter"))))
    (should (git-gutter:in-git-repository-p))))

(ert-deftest git-gutter ()
  "Should return nil if buffer does not related with file or file is not existed"
  (with-current-buffer (get-buffer-create "*not-related-file*")
    (should-not (git-gutter)))
  (let ((buf (find-file-noselect "not-found")))
    (with-current-buffer buf
      (should-not (git-gutter)))))

(ert-deftest git-gutter:collect-deleted-line ()
  "Should return lines which start with '-'"
  (let* ((input (mapconcat 'identity
                           (list "-apple" "-melon" "+orange")
                           "\n"))
         (got (git-gutter:collect-deleted-line input)))
    (should (equal got '("apple" "melon")))))

(ert-deftest git-gutter:insert-deleted-lines ()
  "Should insert deleted line"
  (let ((input (mapconcat 'identity
                          (list "-apple" "-melon" "+orange")
                          "\n")))
    (with-temp-buffer
      (git-gutter:insert-deleted-lines input)
      (should (string= (buffer-string)
                       "apple\nmelon\n")))))

(ert-deftest git-gutter:diff-content ()
  "Should return diff hunk"
  (let* ((input "@@-1,1+1,1@@
foo
bar
@@ -2,2 +2,2 @@")
         (got (with-temp-buffer
                (insert input)
                (goto-char (point-min))
                (goto-char (line-end-position))
                (git-gutter:diff-content))))
    (should (string= got "@@-1,1+1,1@@\nfoo\nbar"))))

(ert-deftest git-gutter:set-window-margin ()
  "Should change window margin"
  (git-gutter:set-window-margin 4)
  (let ((got (car (window-margins))))
    (should (= got 4))))

(ert-deftest git-gutter-mode-success ()
  "Case git-gutter-mode enabled"
  (with-current-buffer (find-file-noselect "test-git-gutter.el")
    (git-gutter-mode 1)
    (should git-gutter-mode))
  (kill-buffer "test-git-gutter.el"))

(ert-deftest git-gutter-mode-failed ()
  "Case git-gutter-mode disabled"
  (with-temp-buffer
    (git-gutter-mode 1)
    (should-not git-gutter-mode))

  (let ((default-directory nil))
    (git-gutter-mode 1)
    (should-not git-gutter-mode))

  (let ((default-directory "foo"))
    (git-gutter-mode 1)
    (should-not git-gutter-mode))

  (when (file-directory-p ".git") ;; #36
    (with-current-buffer (find-file-noselect ".git/config")
      (git-gutter-mode 1)
      (should-not git-gutter-mode))))

(ert-deftest global-git-gutter-mode-success ()
  "Case global-git-gutter-mode enabled"
  (with-current-buffer (find-file-noselect "test-git-gutter.el")
    (global-git-gutter-mode t)
    (should git-gutter-mode))

  (kill-buffer "test-git-gutter.el"))

(ert-deftest global-git-gutter-mode-failed ()
  "Case global-git-gutter-mode disabled"

  (with-temp-buffer
    (global-git-gutter-mode t)
    (should-not git-gutter-mode))

  (let ((git-gutter:disabled-modes '(emacs-lisp-mode)))
    (with-current-buffer (find-file-noselect "test-git-gutter.el")
      (global-git-gutter-mode t)
      (should-not git-gutter-mode)))

  (kill-buffer "test-git-gutter.el"))

(ert-deftest git-gutter-git-diff-arguments ()
  "Command line options of `git diff'"

  (let ((git-gutter:diff-option "-a -b -c")
        (file "git-gutter.el"))
    (let ((got (git-gutter:git-diff-arguments file)))
      (should (equal got '("-a" "-b" "-c" "git-gutter.el"))))

    (let* ((git-gutter:start-revision "HEAD")
           (got (git-gutter:git-diff-arguments file)))
      (should (equal got '("-a" "-b" "-c" "HEAD" "git-gutter.el"))))))

(ert-deftest git-gutter-hg-diff-arguments ()
  "Command line options of `hg diff'"

  (let ((git-gutter:mercurial-diff-option "-a -b -c")
        (file "git-gutter.el"))
    (let ((got (git-gutter:hg-diff-arguments file)))
      (should (equal got '("-a" "-b" "-c" "git-gutter.el"))))

    (let* ((git-gutter:start-revision "30000")
           (got (git-gutter:hg-diff-arguments file)))
      (should (equal got '("-a" "-b" "-c" "-r" "30000" "git-gutter.el"))))))

(ert-deftest git-gutter-bzr-diff-arguments ()
  "Command line options of `bzr diff'"

  (let ((git-gutter:bazaar-diff-option "-a -b -c")
        (file "git-gutter.el"))
    (let ((got (git-gutter:bzr-diff-arguments file)))
      (should (equal got '("-a" "-b" "-c" "git-gutter.el"))))

    (let* ((git-gutter:start-revision "30000")
           (got (git-gutter:bzr-diff-arguments file)))
      (should (equal got '("-a" "-b" "-c" "-r" "30000" "git-gutter.el"))))))

(ert-deftest git-gutter-read-header ()
  "Read header of diff hunk"

  (let ((got (git-gutter:read-hunk-header "@@ -658,31 +688,30 @@")))
    (should (= (nth 0 got) 658))
    (should (= (nth 1 got) 31))
    (should (= (nth 2 got) 688))
    (should (= (nth 3 got) 30)))

  (let ((got (git-gutter:read-hunk-header "@@ -100 +200 @@")))
    (should (= (nth 0 got) 100))
    (should (= (nth 1 got) 1))
    (should (= (nth 2 got) 200))
    (should (= (nth 3 got) 1))))

(ert-deftest git-gutter-show-backends ()
  "Show only handled backends."

  (let ((git-gutter:handled-backends '(bzr))
        (expected "Bzr"))
    (should (string= (git-gutter:show-backends) expected)))

  (let ((git-gutter:handled-backends '(git hg bzr))
        (expected "Git/Hg/Bzr"))
    (should (string= (git-gutter:show-backends) expected))))

;;; test-git-gutter.el end here


test-git-gutter.el
;;; test-git-gutter.el --- Test for git-gutter.el

;; Copyright (C) 2014 by Syohei YOSHIDA

;; Author: Syohei YOSHIDA <syohex@gmail.com>

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'ert)
(require 'git-gutter)

;; suppress log message
(setq git-gutter:verbosity 0)

(ert-deftest git-gutter:root-directory ()
  "helper function `git-gutter:root-directory'"
  (let ((file (concat default-directory "test-git-gutter.el")))
    (when (file-directory-p ".git") ;; https://github.com/syohex/emacs-git-gutter/issues/36
      (let ((expected (expand-file-name default-directory))
            (got (git-gutter:root-directory file)))
        (should (string= expected got)))

      ;; Files in .git/ directory are not version-controled
      (let ((default-directory (concat default-directory ".git/")))
        (should-not (git-gutter:root-directory file))))

    ;; temporary directory maybe be version-controled
    (let ((default-directory temporary-file-directory))
      (should-not (git-gutter:root-directory file)))))

(ert-deftest git-gutter:sign-width ()
  "helper function `git-gutter:sign-width'"
  (let ((got1 (git-gutter:sign-width "a"))
        (got2 (git-gutter:sign-width "0123456789")))
    (should (= got1 1))
    (should (= got2 10))))

(ert-deftest git-gutter:select-face ()
  "helper function `git-gutter:select-face'"
  (cl-loop for (type . expected) in '((added . git-gutter:added)
                                      (modified . git-gutter:modified)
                                      (deleted . git-gutter:deleted))
           do
           (should (eq (git-gutter:select-face type) expected)))
  (should-not (git-gutter:select-face 'not-found)))

(ert-deftest git-gutter:select-sign ()
  "helper function `git-gutter:select-sign'"
  (cl-loop for (type . expected) in '((added . "+") (modified . "=") (deleted . "-"))
           do
           (should (string= (git-gutter:select-sign type) expected)))
  (should-not (git-gutter:select-sign 'not-found)))

(ert-deftest git-gutter:propertized-sign ()
  "helper function `git-gutter:propertized-sign'"
  (should (string= (git-gutter:propertized-sign 'added) "+")))

(ert-deftest git-gutter:changes-to-number ()
  "helper function `git-gutter:changes-to-number'"
  (should (= (git-gutter:changes-to-number "") 1))
  (should (= (git-gutter:changes-to-number "123") 123)))

(ert-deftest git-gutter:make-diffinfo ()
  "helper function `git-gutter:make-diffinfo'"
  (let ((diffinfo1 (git-gutter:make-diffinfo 'added "diff1" 10 20))
        (diffinfo2 (git-gutter:make-diffinfo 'deleted "diff2" 5 nil)))
    (cl-loop for (prop . expected) in '((:type . added)
                                        (:start-line . 10) (:end-line . 20))
             do
             (should (eql (plist-get diffinfo1 prop) expected)))
    (should (string= (plist-get diffinfo1 :content) "diff1"))
    (cl-loop for (prop . expected) in '((:type . deleted)
                                        (:start-line . 5) (:end-line . nil))
             do
             (should (eql (plist-get diffinfo2 prop) expected)))
    (should (string= (plist-get diffinfo2 :content) "diff2"))))

(ert-deftest git-gutter:in-git-repository-p ()
  "Should return nil if default-directory does not exist"

  ;; In git repository, but here is '.git'
  (let ((file (concat default-directory "test-git-gutter.el")))
    (when (file-directory-p ".git") ;; #36
      (let ((buf (find-file-noselect ".git/config")))
        (with-current-buffer buf
          (should-not (git-gutter:in-git-repository-p file)))))

    (let ((default-directory (file-name-directory (locate-library "git-gutter"))))
      (should (git-gutter:in-git-repository-p file)))))

(ert-deftest git-gutter ()
  "Should return nil if buffer does not related with file or file is not existed"
  (with-current-buffer (get-buffer-create "*not-related-file*")
    (should-not (git-gutter)))
  (let ((buf (find-file-noselect "not-found")))
    (with-current-buffer buf
      (should-not (git-gutter)))))

(ert-deftest git-gutter:collect-deleted-line ()
  "Should return lines which start with '-'"
  (let* ((input (mapconcat 'identity
                           (list "-apple" "-melon" "+orange")
                           "\n"))
         (got (git-gutter:collect-deleted-line input)))
    (should (equal got '("apple" "melon")))))

(ert-deftest git-gutter:insert-deleted-lines ()
  "Should insert deleted line"
  (let ((input (mapconcat 'identity
                          (list "-apple" "-melon" "+orange")
                          "\n")))
    (with-temp-buffer
      (git-gutter:insert-deleted-lines input)
      (should (string= (buffer-string)
                       "apple\nmelon\n")))))

(ert-deftest git-gutter:diff-content ()
  "Should return diff hunk"
  (let* ((input "@@-1,1+1,1@@
foo
bar
@@ -2,2 +2,2 @@")
         (got (with-temp-buffer
                (insert input)
                (goto-char (point-min))
                (goto-char (line-end-position))
                (git-gutter:diff-content))))
    (should (string= got "@@-1,1+1,1@@\nfoo\nbar"))))

(ert-deftest git-gutter:set-window-margin ()
  "Should change window margin"
  (git-gutter:set-window-margin 4)
  (let ((got (car (window-margins))))
    (should (= got 4))))

(ert-deftest git-gutter:file-path ()
  "Should return file path which is passed to 'git diff'"
  (let* ((file (concat default-directory "test-git-gutter.el"))
         (expected file)
         (got (git-gutter:file-path default-directory file)))
    (should (string= got expected))))

(ert-deftest git-gutter-mode-success ()
  "Case git-gutter-mode enabled"
  (with-current-buffer (find-file-noselect "test-git-gutter.el")
    (git-gutter-mode 1)
    (should git-gutter-mode))
  (kill-buffer "test-git-gutter.el"))

(ert-deftest git-gutter-mode-failed ()
  "Case git-gutter-mode disabled"
  (with-temp-buffer
    (git-gutter-mode 1)
    (should-not git-gutter-mode))

  (let ((default-directory nil))
    (git-gutter-mode 1)
    (should-not git-gutter-mode))

  (let ((default-directory "foo"))
    (git-gutter-mode 1)
    (should-not git-gutter-mode))

  (when (file-directory-p ".git") ;; #36
    (with-current-buffer (find-file-noselect ".git/config")
      (git-gutter-mode 1)
      (should-not git-gutter-mode))))

(ert-deftest global-git-gutter-mode-success ()
  "Case global-git-gutter-mode enabled"
  (with-current-buffer (find-file-noselect "test-git-gutter.el")
    (global-git-gutter-mode t)
    (should git-gutter-mode))

  (kill-buffer "test-git-gutter.el"))

(ert-deftest global-git-gutter-mode-failed ()
  "Case global-git-gutter-mode disabled"

  (with-temp-buffer
    (global-git-gutter-mode t)
    (should-not git-gutter-mode))

  (let ((git-gutter:disabled-modes '(emacs-lisp-mode)))
    (with-current-buffer (find-file-noselect "test-git-gutter.el")
      (global-git-gutter-mode t)
      (should-not git-gutter-mode)))

  (kill-buffer "test-git-gutter.el"))

;;; test-git-gutter.el end here
